In [528]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [529]:
# Load the dataset (load remotely or locally)

iris_data = pd.read_csv('iris.csv')

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

print(iris_data[0:15])
iris_data.info()

    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

This data is on the relationship between sepal and petal sizes with the plant species. The features are the sepal_length, sepal_width, petal_length, and petal_width. The label is the species column. The labels setosa, versicolor, and virginica directly map to the three classes of flowers shown in the readme. 

# Part 2: Split the dataset into train and test

In [530]:
# Take the dataset and split it into our features (X) and label (y)
feature_cols = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width']
features = iris_data[feature_cols]
label = iris_data['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(features, label, test_size=0.1, random_state=42)


# Part 3: Logistic Regression

In [531]:
# i. Use sklearn to train a LogisticRegression model on the training set

regression = LogisticRegression(random_state=42).fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

sample_data_point = pd.DataFrame([[4.0, 4.0, 3.0, 1.0]], columns=feature_cols)
print(f'Class probability order: {regression.classes_}')
print(f'Probabilities: {regression.predict_proba(sample_data_point)}')

# iii. Report on the score for Logistic regression model, what does the score measure?

print(f'Score: {regression.score(x_test, y_test)}')

# iv. Extract the coefficents and intercepts for the boundary line(s)
print(f'Intercepts: {regression.intercept_}')
print(f'Coefficients: {regression.coef_}')

for i in range(0, 3):
    print(f'Boundary for {regression.classes_[i]}:')
    print(f'{regression.intercept_[i]}', end='')
    for j in range(0, 4):
        print(f' + {regression.coef_[i][j]} * x_{j+1}')

    print()

Class probability order: ['setosa' 'versicolor' 'virginica']
Probabilities: [[8.50075986e-01 1.49694149e-01 2.29864854e-04]]
Score: 1.0
Intercepts: [  9.52941208   1.897714   -11.42712608]
Coefficients: [[-0.42612585  0.96865398 -2.44812859 -1.03769847]
 [ 0.51149531 -0.21830267 -0.21574601 -0.8425377 ]
 [-0.08536945 -0.75035131  2.6638746   1.88023617]]
Boundary for setosa:
9.529412084058322 + -0.42612585150396853 * x_1
 + 0.9686539759227968 * x_2
 + -2.44812859074156 * x_3
 + -1.0376984724624452 * x_4

Boundary for versicolor:
1.8977139981293287 + 0.5114953064385493 * x_1
 + -0.21830267079944407 * x_2
 + -0.21574601313415478 * x_3
 + -0.8425377003870715 * x_4

Boundary for virginica:
-11.427126082187879 + -0.0853694549345783 * x_1
 + -0.7503513051233466 * x_2
 + 2.6638746038757253 * x_3
 + 1.8802361728495203 * x_4



iii. The score represents the average of the accuracy for each of the datapoints in the test data. In this case, we can see that on average, the model correctly classified the test data points 100% of the time.

# Part 4: Support Vector Machine

In [532]:
# i. Use sklearn to train a Support Vector Classifier on the training set

vector_classifier = SVC(probability=True, random_state=42).fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

print(f'SVC Class Order: {vector_classifier.classes_}')
print(f'SVC Probabilities: {vector_classifier.predict_proba(sample_data_point)}')

# iii. Report on the score for the SVM, what does the score measure?
print(f'Score: {vector_classifier.score(x_test, y_test)}')

SVC Class Order: ['setosa' 'versicolor' 'virginica']
SVC Probabilities: [[0.57546609 0.3613088  0.0632251 ]]
Score: 1.0


iii. The score again measures the average accuracy over all of the datapoints in the test set. In this case, it was again 100% accurate at classifying each of the test data points.

# Part 5: Neural Network

In [533]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set

mlp_classifier = MLPClassifier(random_state=42, max_iter=500).fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

print(f'Class Order: {mlp_classifier.classes_}')
print(f'Probabilities: {mlp_classifier.predict_proba(sample_data_point)}')

# iii. Report on the score for the Neural Network, what does the score measure?

print(f'Score: {mlp_classifier.score(x_test, y_test)}')

# iv: Experiment with different options for the neural network, report on your best configuration
print(f'10 hidden layers (instead of 100): {MLPClassifier(hidden_layer_sizes=10, max_iter=500, random_state=42).fit(x_train, y_train).score(x_test, y_test)}')
print(f'lbfgs instead of gradient descent: {MLPClassifier(solver='lbfgs', max_iter=500, random_state=42).fit(x_train, y_train).score(x_test, y_test)}')
print(f'High learning rate alpha=1: {MLPClassifier(alpha=1, max_iter=500, random_state=42).fit(x_train, y_train).score(x_test, y_test)}')

Class Order: ['setosa' 'versicolor' 'virginica']
Probabilities: [[6.40488855e-01 3.59504798e-01 6.34684907e-06]]
Score: 1.0
10 hidden layers (instead of 100): 0.8666666666666667
lbfgs instead of gradient descent: 1.0
High learning rate alpha=1: 1.0


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


iv: Out of the four configurations I tried, the only one that was not completely accurate on the test data was the MLP with only 10 hidden layers. That model only predicted the test data on average 86.67% of the time.

# Part 6: K-Nearest Neighbors

In [534]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

kn_classifier = KNeighborsClassifier().fit(x_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
print(f'Class Order: {kn_classifier.classes_}')
print(f'Probabilities: {kn_classifier.predict_proba(sample_data_point)}')

# iii. Report on the score for kNN, what does the score measure?
print(f'Score: {kn_classifier.score(x_test, y_test)}')

Class Order: ['setosa' 'versicolor' 'virginica']
Probabilities: [[0.6 0.4 0. ]]
Score: 1.0


iii. Once again, the score measures the average accuracy for the model's predictions on the test data. It was once again 100%.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Out of all of the models, the only one that didn't have a 100% test accuracy was the 10 hidden layer MLP. Therefore, the rest performed equally on the given dataset, since they correctly predicted every test data point. I think this indicates that the three classes from the iris dataset must be fairly distinct from eachother, since the models are able to make accurate predictions with so few data points. Going into this, I would've expected much more variation in the scores for the models. I was especially surprised that KNN performed as well as the other models, given how simple the algorithm is.